Source/WebCore/editing/cocoa/TextIteratorHTMLConverter.mm

```diff
@@ -0,0 +1,357 @@
+
+        return hasAncestorQualifyingForWritingToolsPreservation(ancestor->parentElement(), cache);
+    }).iterator->value;
+}
+
+#pragma mark - EditingHTMLConverter definition.
+
+class EditingHTMLConverter {
+public:
+    explicit EditingHTMLConverter(const SimpleRange&, OptionSet<IncludedElement>);
+    ~EditingHTMLConverter();
+
+    AttributedString convert();
+
+private:
+    SimpleRange m_range;
+    OptionSet<IncludedElement> m_includedElements;
+};
+
+#pragma mark - EditingHTMLConverter implementation.
+
+EditingHTMLConverter::EditingHTMLConverter(const SimpleRange& range, OptionSet<IncludedElement> includedElements)
+    : m_range(range)
+    , m_includedElements(includedElements)
+{
+}
+
+EditingHTMLConverter::~EditingHTMLConverter() = default;
+
+// This function uses TextIterator, which makes offsets in its result compatible with HTML editing.
+AttributedString EditingHTMLConverter::convert()
+{
+#if PLATFORM(MAC)
+    auto fontManager = [NSFontManager sharedFontManager];
+#endif
+
+    HashMap<Element*, bool> m_elementQualifiesForWritingToolsPreservationCache;
+
+    auto string = adoptNS([[NSMutableAttributedString alloc] init]);
+    auto attrs = adoptNS([[NSMutableDictionary alloc] init]);
+    NSUInteger stringLength = 0;
+    for (TextIterator it(m_range); !it.atEnd(); it.advance()) {
+        auto node = it.node();
+
+        if (RefPtr imageElement = dynamicDowncast<HTMLImageElement>(node); imageElement && m_includedElements.contains(IncludedElement::Images)) {
+            RetainPtr attachmentAttributedString = attributedStringWithAttachmentForElement(*imageElement);
+            [string appendAttributedString:attachmentAttributedString.get()];
+            stringLength += [attachmentAttributedString length];
+        }
+
+        if (RefPtr attachmentElement = dynamicDowncast<HTMLAttachmentElement>(node); attachmentElement && m_includedElements.contains(IncludedElement::Attachments)) {
+            RetainPtr attachmentAttributedString = attributedStringWithAttachmentForElement(*attachmentElement);
+            [string appendAttributedString:attachmentAttributedString.get()];
+            stringLength += [attachmentAttributedString length];
+        }
+
+        auto currentTextLength = it.text().length();
+        if (!currentTextLength)
+            continue;
+
+        // In some cases the text iterator emits text that is not associated with a node.
+        // In those cases, base the style on the container.
+        if (!node)
+            node = it.range().start.container.ptr();
+        auto renderer = node->renderer();
+        ASSERT(renderer);
+        if (!renderer)
+            continue;
+        auto& style = renderer->style();
+
+#if ENABLE(WRITING_TOOLS)
+        // FIXME: Use the `WTWritingToolsPreservedAttributeName` constant instead of a string.
+        if (hasAncestorQualifyingForWritingToolsPreservation(node->parentElement(), m_elementQualifiesForWritingToolsPreservationCache))
```

Comment:
We should add an option used when creating `EditingHTMLConverter`, since this logic shouldn't be run when generating attributed strings for other purposes.